src/jobserver.h

```diff
@@ -0,0 +1,224 @@
+
+    /// Copy operations are disallowed.
+    Slot(const Slot&) = delete;
+    Slot& operator=(const Slot&) = delete;
+
+    /// Return true if this instance is valid, i.e. that it is either
+    /// implicit or explicit job slot.
+    bool IsValid() const { return value_ >= 0; }
+
+    /// Return true if this instance represents an implicit job slot.
+    bool IsImplicit() const { return value_ == kImplicitValue; }
+
+    /// Return true if this instance represents an explicit job slot
+    bool IsExplicit() const { return IsValid() && !IsImplicit(); }
+
+    /// Return value of an explicit slot. It is a runtime error to call
+    /// this from an invalid instance.
+    uint8_t GetExplicitValue() const;
+
+    /// Create instance for explicit byte value.
+    static Slot CreateExplicit(uint8_t value) {
+      return Slot(static_cast<int16_t>(value));
+    }
+
+    /// Create instance for the implicit value.
+    static Slot CreateImplicit() { return Slot(kImplicitValue); }
+
+   private:
+    Slot(int16_t value) : value_(value) {}
+
+    static constexpr int16_t kImplicitValue = 256;
+
+    int16_t value_ = -1;
+  };
+
+  /// A Jobserver::Config models how to access or implement a GNU jobserver
+  /// implementation.
+  struct Config {
+    /// Different implementation modes for the slot pool.
+    ///
+    /// kModeNone means there is no pool.
+    ///
+    /// kModePipe means that `--jobserver-auth=R,W` is used to
+    ///    pass a pair of file descriptors to client processes. This also
+    ///    matches `--jobserver-fds=R,W` which is an old undocumented
+    ///    variant of the same scheme. This mode is not supported by
+    ///    Ninja, but recognized by the parser.
+    ///
+    /// kModePosixFifo means that `--jobserver-auth=fifo:PATH` is used to
+    ///    pass the path of a Posix FIFO to client processes. This is not
+    ///    supported on Windows. Implemented by GNU Make 4.4 and above
+    ///    when `--jobserver-style=fifo` is used.
+    ///
+    /// kModeWin32Semaphore means that `--jobserver-auth=SEMAPHORE_NAME` is
+    ///    used to pass the name of a Win32 semaphore to client processes.
+    ///    This is not supported on Posix.
+    ///
+    /// kModeDefault is the default mode to enable on the current platform.
+    ///    This is an alias for kModeWin32Semaphore on Windows ,and
+    ///    kModePosixFifo on Posix.
+    enum Mode {
+      kModeNone = 0,
+      kModePipe,
+      kModePosixFifo,
+      kModeWin32Semaphore,
+#ifdef _WIN32
+      kModeDefault = kModeWin32Semaphore,
+#else   // _WIN32
+      kModeDefault = kModePosixFifo,
+#endif  // _WIN32
+    };
+
+    /// Implementation mode for the pool.
+    Mode mode = kModeNone;
+
+    /// For kModeFifo, this is the path to the Unix FIFO to use.
+    /// For kModeSemaphore, this is the name of the Win32 semaphore to use.
+    std::string path;
+
+    /// For kModePipe, these are the file descriptor values
+    /// extracted from MAKEFLAGS.
+    int read_fd = -1;
```

Comment:
Can be removed now and turned into local variables inside GetFileDescriptorPair.